src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Rpc
+{
+    /// <summary>
+    /// Implements the server side of the RPC channel used to communicate with the build host.
+    /// </summary>
+    /// <remarks>
+    /// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+    /// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+    /// are out. If at some point there is a standard RPC mechanism exposed in .NET, we should just delete this and use that instead.
+    /// </remarks>
+    internal sealed class RpcServer
+    {
+        private readonly TextWriter _sendingStream;
+        private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+        private readonly TextReader _receivingStream;
+
+        private readonly ConcurrentDictionary<int, object> _rpcTargets = new();
+        private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
```

Comment:
don't think this needs to be volaite if you're using Interlocked on it.  @stephentoub ?

Reply:
It looks like the only use is in an Interlocked.Increment? In which case, correct, the volatile here is not needed (nor will it have any impact).